### DISCUSSION TITLE
Easier/general way to implement Typescript's "key remapping via `as`" approach

### DISCUSSION THREAD

**Discussion by jrencz** [sinclairzx81/typebox#985] [Ideas]
What I refer to in the title as 'the "key remapping via `as`" approach' is https://www.typescriptlang.org/docs/handbook/2/mapped-types.html#key-remapping-via-as

Looking through resolved issues related to Template Literal and Mapped types I found the resolved #913 

Inspired by that I tried to rebuild the example from the TS Handbook, (link below) and I think it shows an idea for a nice feature to be added: a way to use both the `TMappedKey` and the original key in `TMappedFunction`.

I imagine it would mean `Type.Mapped` knows not only about the "new" keys (as it does now, the 1st argument) but somehow it can be given both the keys and the function that **maps keys**, not only the function that **maps values**, the `TMappedFunction`.

Here's [playground](https://www.typescriptlang.org/play/?ts=4.9.5#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxjsriQBYYgFwC+cAZlBCHAOQACAzsADsSAGwLAoAehjoMAIwgAPTgChlEiXABibDlRgwwvAFzqA7uYB00zLxJRgYGKIEBzC9BcSAJhBK8JVAQCXvIQANYSAEwSdGCYXgC01hi8FnogwgDEYRioCVC0BHGCLgkAbsAECQS8yiQQArzwAAoYULwNcAC8KDIWAPKyAFYYJDAAFAjKcHACBCAYRr2YFvj2ruMAlFjTcAQui8sYFgByAK4gsm1bOzPCvkTADUtoK2slN8oMmwDcdQ1NOBDCCHfCPEgAHmSEGYcFa7QaAD5uog5gslpwAFIgzg4faHSIARhw9xIj2eXAEEFMNAKuIYqnqjXgABkCAAvVDwjoCMEHQko17HACyRXi412QosAGEisBCMJgOyMOMpQBpXL9ZjjbkNTbbXbqOAnCAwQ5BOAQWRNAgiDBwBKWgT2wK8OAwCBwQJle30Ao+9r2mBQIK8ZjQEB4t1quDAN30WLxOA5VCGjQAA2SKFFcQwXk0ZyExAaEJjGAUZuCbuabEwsFQGtQAG0ALo4ACSgpz8UbpcRyJ6qqW7c23WRFGohR+6bTcA6LAIUFjME4bowNVQOHlq7gYDYFS8ed2Ma64767eC5Z1bR55D6AFUhHKFUqVWr9cpfv9mXBRJzsRgfIYISSxgsQkLQrCbKcrqvKEPyA6ILsJzzIcWI4rccAAIIHEsRKYSyDzFgIGJUjSbQYLiXyqEaQFwISFgAKwOnAZy8EG9gVAQwgoLQYCiGaLLym03Gxsy65eGoGgmmaw7wA0wioHApjQGEbqCCguAACwWAAnAA1CxNJKXG7qBDAAD8cJsLIBCyIpsymnsAixrwvBnPa-BCPaOm6UxrlmfaEDCIegLuXE0Bml4FjfoC0FcjeDRAYSAD6zE9FK3Z5hKMxSsgfECRgQlmiGwjjE2uy5X0spgPK3GvqqfSNlq14IgI+qYS2BozCeZ4rBeh4KK1t5HBYj5krVL7KuM76bJ+fxMoCf4JW1yVpaBhDgVCMgwnA8WwWtjGIVMMwoeiXAAVRMw4QSxK7IRZLEaR1K0pROwMlJeDwfakQsXQORwGU3EeXsLCFmMTwCLFrIcitPJAb9GV9FlXg5aN+XgIVxUiWVFUzFVKw1XVirTeqmrarBHW7F1mG9aNBZFpD5VtqNA1XrBd4rONz71dNs1zV+i3wMtB3fZEG3gttmC7ftiVwfikTHchqFLFsY7oW9uw3aro6nnA+H3URkM62OlIvRRVEfbR31wAAzCxBBeF4u4FMwwAKO6nowDQya5HA4zMGcwjCAkzDrjAZwFM7jZwAASoUxSuIDlRwOmNTpg76kud79oABJBCEEDhHNQt7bDov4vbSMrCjaN5QVRBFcJpXlZVo3Yy3nAHCu3X41KRNTSqZOoC1lO9zTs62SQYSe4FLsYBUEBsQ5CwCMReaWtahB2ixOcFF6NSz1coP+jeHGhuGUB0MRlqwjn89uwoR4zEaioldxFhwNzk28-aplWovZeSkqTwAwFSM4LgqCf1nOmKUbMhoc1Gs1Cmct9QZ1MkDFwHkcB7ggF4M4JAUipyaOsFwcAAA+swLhXCgBnR0IDlLmR3M6PMz84DqFnKKVAx9txukgG5YA9kgyenLMGW08B772EgRIqgJRb5zgIVQUG-BMZBgbmaX8zduI4FkGcGRRBmGmlnOIs43EHL30gIIGAaZZztjvq9ZSh8LRLxgAkGESQaAJHkB7UwBAlIejgOWfh9oABC-RkC5znomDeBdWICGic7FMYlsxijzAzCGDRZzymUkvEKcAFgWjTrwDOvBUBrwIB7XaQpbD2EcDuBJ7oZBuhUkHZ2x96gBg3uMIgv51yAjAIuYgwdUCjgLlk+ALS8nHz6f4z0bFjiziwjI+0IB2BgOWU6ByvilLsVoG6AJx974ACJT6BndCGRol9r6QyOXsOIbBbRKMPhpOikRjymylOk4izNObHHgcNBovyxpPh-iTN8+oBZ-CzLLVa30q5fUlpBMuME5ZAVtoiaGv5y6orhUsGF8M4UohOnAbuZ1Dhqz1hrK6JKMAwG1v7XWyIDYzG7g9ckJEGWm04GRV6ltVBAA) with description of the concept, until the point, where the suggested new syntax would make things easier.

**Comment by sinclairzx81** [comment]
@jrencz Hi, thanks for the suggestion! This is a cool idea.

I'll spend some time looking into this when I next get a bit of free time to deep dive into this aspect of the library. I think being able to remap property names would be awesome (an essential) addition to the Mapped type feature. I expect this feature would need to be realized by way of new type (`Type.MappedAs`, `Type.KeyAs`, `Type.As` or similar), something to provide the Mapped functions key parameter a bit more context (essentially a source and target mapping for each property key)

There's not too much I can implement at this stage, so will convert this issue into a discussion thread for now as new features like this are usually best discussed there. Happy to field any questions if you wanted to have a go at a prototype implementation :)

Cheers!
S

**Reply by jrencz** [reply]
> I expect this feature would need to be realized by way of new type (`Type.MappedAs`, `Type.KeyAs`, `Type.As` or similar), something to provide the Mapped functions key parameter a bit more context (essentially a source and target mapping for each property key)

Just purely from the user's point of view, I think having a distinct function to create this flavour of mapped type would be less predictable than having some kind of yet another a signature of the `Type.Mapped` but for those already existing:

https://github.com/sinclairzx81/typebox/blob/d387b2adb7532d7059d5be9cffe9bb8552d72566/src/type/mapped/mapped.ts#L261-L264

```ts
export function Mapped<…>(key: K, keyMap: F1, map: F2, options?: ObjectOptions): R
```
(I'm skipping the type parameters for brevity)

Here's why I think it would make sense to at least give it a try to fit the `as` syntax into the `Type.Mapped` with 3(4 with `options`) arguments (where I honestly don't have experience to judge if it makes sense or if it's going to make implementation harder, but anyway): **3 elements is what AST for that original syntax is**.

Let's consider the AST Typescript produces for both syntaxes:

```ts
type NoRemap<Type> = {
    [Property in keyof Type]: Type[Property]
};

type WithRemap<Type> = {
    [Property in keyof Type as `prefixed${string & Property}`]: Type[Property]
};
```

https://ts-ast-viewer.com/#code/C4TwDgpgBAcg9gJQgWwIZgDwBVwQHxQC8UA3gFBSVQDaACgE5yT2hQCWAdlANYQhwAzKDkgBdAFzDcdRs1CiyAXwDcZMqEhQA6m2AALJGkwj8RUhSoymEFiHZde-ISaioAzlAAGYehAFsADwgAEwASEjdgek4AcygAMigGa1tFTwkpSCs5EAUVMiA

```
SourceFile
  TypeAliasDeclaration
    Identifier
    TypeParameter
    MappedType
      TypeParameter     <-- in Typebox it's the TKeyOf
      IndexedAccessType <-- in Typebox it's the TMappedFunction
  TypeAliasDeclaration
    Identifier
    TypeParameter
    MappedType
      TypeParameter
      TemplateLiteralType <-- in Typebox it would be a TMappedKeyMapFunction (proposed name)
      IndexedAccessType
  EndOfFileToken
```

in Typebox the first is

```ts
Type.Mapped(
  Type.KeyOf(T),

  // K is a TKeyOf
  (K) => Type.Index(Type, K)
)
```

the second I propose to be

```ts
Type.Mapped(
  Type.KeyOf(Type),

  // what I earlier named TMappedKeyMapFunction
  // K here is a TKeyOf
  (K) => Type.TemplateLiteral([Type.Const('prefixed'), K])
  
  // in proposed 3(4)-argument signature TMappedFunction would have not 1 but 2 arguments
  // or maybe there's a TMappedAsFunction (another proposed name) type, in which where
  // - K is a TKeyOf - same as in TMappedFunction
  // - A is the TTemplateLiteral - possibly for convenience of nor repeating oneself, in case it
  //   actually might be needed (it's not used in discussed case)
  (K, A) => Type.Index(Type, K)
)
```

**Reply by jrencz** [reply]
may I ask if you had a chance to think about it yet?
